api/cluster/controller.go

```diff
@@ -325,28 +328,107 @@ func (c *controller) deleteInferenceService(serviceName string, namespace string
 	return nil
 }
 
-func (c *controller) waitInferenceServiceReady(service *kservev1beta1.InferenceService) (*kservev1beta1.InferenceService, error) {
+func (c *controller) waitInferenceServiceReady(service *kservev1beta1.InferenceService) (isvc *kservev1beta1.InferenceService, err error) {
+	ctx := context.Background()
+
 	timeout := time.After(c.deploymentConfig.DeploymentTimeout)
-	ticker := time.NewTicker(time.Second * tickDurationSecond)
+
+	isvcConditionTable := ""
+	podContainerTable := ""
+	podLastTerminationMessage := ""
+
+	defer func() {
+		if err == nil {
+			return
+		}
+
+		if isvcConditionTable != "" {
+			err = fmt.Errorf("%w\n\nModel service conditions:\n%s", err, isvcConditionTable)
+		}
+
+		if podContainerTable != "" {
+			err = fmt.Errorf("%w\n\nPod container status:\n%s", err, podContainerTable)
+		}
+
+		if podLastTerminationMessage != "" {
+			err = fmt.Errorf("%w\n\nPod last termination message:\n%s", err, podLastTerminationMessage)
+		}
+	}()
+
+	isvcWatcher, err := c.kserveClient.InferenceServices(service.Namespace).Watch(metav1.ListOptions{
+		FieldSelector: fmt.Sprintf("metadata.name=%s", service.Name),
+	})
+	if err != nil {
+		return nil, errors.Wrapf(err, "unable to initialize isvc watcher: %s", service.Name)
+	}
+
+	podWatcher, err := c.clusterClient.Pods(service.Namespace).Watch(ctx, metav1.ListOptions{
+		LabelSelector: fmt.Sprintf("serving.kserve.io/inferenceservice=%s", service.Name),
+	})
+	if err != nil {
+		return nil, errors.Wrapf(err, "unable to initialize isvc's pods watcher: %s", service.Name)
+	}
 
 	for {
 		select {
 		case <-timeout:
-			log.Errorf("timeout waiting for inference service to be ready %s", service.Name)
 			return nil, ErrTimeoutCreateInferenceService
-		case <-ticker.C:
-			s, err := c.kserveClient.InferenceServices(service.Namespace).Get(service.Name, metav1.GetOptions{})
-			if err != nil {
-				log.Errorf("unable to get inference service status %s %v", service.Name, err)
-				return nil, ErrUnableToGetInferenceServiceStatus
+
+		case isvcEvent := <-isvcWatcher.ResultChan():
+			isvc, ok := isvcEvent.Object.(*kservev1beta1.InferenceService)
+			if !ok {
+				return nil, errors.New("unable to cast isvc object")
 			}
+			log.Debugf("isvc event received [%s]", isvc.Name)
 
-			if s.Status.IsReady() {
-				// Inference service is completely ready
-				return s, nil
+			if isvc.Status.Status.Conditions != nil {
+				isvcConditionTable, err = parseInferenceServiceConditions(isvc.Status.Status.Conditions)
```

Comment:
would it be better to keep track of the last state of  `isvc, ok := isvcEvent.Object.(*kservev1beta1.InferenceService)` and `pod, ok := podEvent.Object.(*corev1.Pod)`, then do the parsing once in the defer instead of every message

Reply:
In my opinion, in terms of performance, your suggestion would be more efficient since we don't need to parse the k8s objects on every event. But I prefer to have `defer` function as short as possible, hence no need to have complex logic here and just return the error message if exist.